### PR TITLE
Chore/enable rust backtrace in tests

### PR DESCRIPTION
**Pull request by casimiro**

### CODE REVIEW

**Comment by thibaultcha** [comment]
I don't really understand the purpose of this. We have the `backtraces` directive to enable detailed backtraces, we shouldn't directly play with environment variables as each runtimes (and underlying SDK whether Rust or Go or other) has different ways of enabling detailed backtraces. If a test is not giving a detailed backtraces and needs individual investigation, it takes very little effort to isolate it and add whatever is necessary to extract more details, whether it be a backtrace or something else.

**Comment by thibaultcha** [comment]
Besides, adding `backtraces on` by default to all tests would also defeat the purpose, since it currently allows us to test both `backtraces [on|off]` throughout the whole test suite (`off` by default and `on` in `backtraces_directive.t`), which has been by design.

File: t/TestWasm.pm [hunk 1]
```diff
@@ -108,9 +108,13 @@ add_block_preprocessor(sub {
...
-    $block->set_value("main_config",
-                      "env WASMTIME_BACKTRACE_DETAILS=1;\n"
-                      . $main_config);
+    my @env_vars = (
+        "env WASMTIME_BACKTRACE_DETAILS=1",
+        "env RUST_BACKTRACE=full",
+    );
+    my $env_section = join(";\n", @env_vars) . ";\n\n";
+
+    $block->set_value("main_config", $env_section . $main_config);
```
Comment (thibaultcha) [inline t/TestWasm.pm:117]: Maybe https://metacpan.org/pod/Test::Nginx::Socket#env_to_nginx would be better suited for this task, even though for now I'm more focused on the other points made in the PR thread.

File: t/TestWasm.pm [hunk 2]
```diff
@@ -108,9 +108,13 @@ add_block_preprocessor(sub {
 
     my $main_config = $block->main_config || '';
 
-    $block->set_value("main_config",
-                      "env WASMTIME_BACKTRACE_DETAILS=1;\n"
```
Comment (thibaultcha) [inline t/TestWasm.pm:112]: It just struck me this is still set even though we have now added the `backtraces` directive... It should have probably been removed at the time.

**Comment by thibaultcha** [comment]
What do you think about:

1. Leaving `stub` as-is (granted it doesn't cause CI failures with longer backtraces names etc...)
2. Unconditionally add `RUST_BACKTRACE` through `setenv` as part of `ngx_wrt_wasmtime.c`
3. Removing `WASMTIME_BACKTRACE_DETAILS` from `TestWasm.pm`

File: t/01-wasm/directives/003-shm_directives.t [hunk 3]
```diff
@@ -22,7 +22,7 @@ __DATA__
 [error]
 [crit]
 [emerg]
-stub
+[stub]
```
[the remark this answers is not in the document]
Comment (casimiro) [reply t/01-wasm/directives/003-shm_directives.t:25]: Oh, my bad. I noticed `[stubN]` being used in `t/03-proxy_wasm/hfuncs/shm/004-kv_eviction.t` then I assumed it was an emerging pattern we hadn't yet propagated to tests previously edited.

**Comment by casimiro** [comment]
Thank you for your careful review @thibaultcha!

To be fair, what motivated these commits was having a convenient way to get Rust backtraces to the logs when investigating an erratic behaviour involving Wasmtime or Wasmer.

As you said, though, it can be easily achieved by adding `backtraces on;` to the test case under investigation.

I agree with the 3 suggestions, but maybe we only set `RUST_BACKTRACE` using `setenv` if `backtraces on;` is present?
I think it's safe to assume that when users set `backtraces on;` they expect to have any backtraces available.

That way we can have `RUST_BACKTRACE` set only when Wasmtime or Wasmer is used, following our current approach to enable `WASMTIME_BACKTRACE_DETAILS` [here](https://github.com/Kong/ngx_wasm_module/blob/main/src/wasm/wrt/ngx_wrt_wasmtime.c#L136-L141).

File: t/01-wasm/directives/003-shm_directives.t (hunk 3 above)
[the remark this answers is not in the document]
Comment (thibaultcha) [reply t/01-wasm/directives/003-shm_directives.t:25]: Oh yeah... Well, we have `[stub]` be the norm too if you want to leave those changes, not a big deal.

**Comment by thibaultcha** [comment]
With all that, it could indeed be interesting to add a `--- backtraces` block to TestWasm.pm as you were doing in the previous PR! As a way to easily add `backtraces on;` to any test case without having to add a whole `--- main_config` block.

**Pull request closed without merging**